mycpp/control_flow_pass.py

```diff
@@ -24,19 +25,111 @@ class UnsupportedException(Exception):
 
 class Build(ExpressionVisitor[T], StatementVisitor[None]):
 
-    def __init__(self, types: Dict[Expression, Type]):
+    def __init__(self, types: Dict[Expression, Type], virtual, local_vars, imported_names):
 
         self.types = types
-        self.cfgs = {}
+        self.cfgs = collections.defaultdict(pass_state.ControlFlowGraph)
         self.current_statement_id = None
+        self.current_class_name = None
         self.current_func_node = None
         self.loop_stack = []
+        self.virtual = virtual
+        self.local_vars = local_vars
+        self.imported_names = imported_names
+        self.callees = {} # statement object -> SymbolPath of the callee
 
     def current_cfg(self):
         if not self.current_func_node:
             return None
 
-        return self.cfgs.get(split_py_name(self.current_func_node.fullname))
+        return self.cfgs[split_py_name(self.current_func_node.fullname)]
+
+    def resolve_callee(self, o: CallExpr) -> Optional[util.SymbolPath]:
+        """
+        Returns the fully qualified name of the callee in the given call
+        expression.
+
+        Member functions are prefixed by the names of the classes that contain
+        them. For example, the name of the callee in the last statement of the
+        snippet below is `module.SomeObject.Foo`.
+
+            x = module.SomeObject()
+            x.Foo()
+
+        Free-functions defined in the local module are referred to by their
+        normal fully qualified names. The function `foo` in a module called
+        `moduleA` would is named `moduleA.foo`. Calls to free-functions defined
+        in imported modules are named the same way.
+        """
+
+        if isinstance(o.callee, NameExpr):
+            return split_py_name(o.callee.fullname)
+
+        elif isinstance(o.callee, MemberExpr):
+            if isinstance(o.callee.expr, NameExpr):
+                is_module = (isinstance(o.callee.expr, NameExpr) and
```

Comment:
I think the use of `self.imported_names` should be eliminated -- that is the main source of duplication I see

We are trying to distinguish the cases

    a.b  # attribute of class/struct  a->b
    a.b()  # module and function a::b()

And the heuristic is to look at if a is an imported name

I think we should figure that out ONCE, and then change the representation for later  passes

e.g. we could introduce a new node type even ...  

Or maybe wrap it it in a new type

At the very least it could save this calculation in a dict for later


----

Logically I think the disambiguation belongs in an EARLY pass.   Though maybe the DECL pass can be the place to stick it.

We really should create something like `mycpp.asdl` that defines the language.

So we know that

    a.b() is different than a.b
    x = 42 is different than x = cast(T, 42)

in terms of what we need to compile

If the early stages do the disambiguation, then the later stages don't have to worry about it, and they have a nicer tree  to work with

---

Logically it does seem like we need this to make a correct control flow graph

Right now the disambiguation part is mixed up with the control flow part.  It was always mixed up but I think this is the right time to fix it -- when we have some significant new functionality that hinges on it